Feature/alvtime v3

boye-s opened the pull request:
This pull request introduces several changes, including the addition of a new `frontend-v3` service, updates to authentication and user handling, and the setup of a new frontend project with supporting configurations. The most significant changes are grouped and summarized below.

### New `frontend-v3` Service
* Added a `frontend-v3` service to `docker-compose.yaml` with its own build context, ports, and volumes. This sets up a development environment for the new frontend project.
* Created a `Dockerfile` for `frontend-v3` with multi-stage builds for development and production, including a non-root user setup and dependency installation.
* Added configuration files for the `frontend-v3` project, including `.devcontainer.json`, `.dockerignore`, `.eslintrc.js`, `.gitignore`, `.prettierrc`, and `.vscode/extensions.json`. These files define development tools, linting rules, and editor settings. [[1]](diffhunk://#diff-9ce3f2698276fa57dd18be9d9c18115e3b405082fe771b44830e040e8ae29aedR1-R17) [[2]](diffhunk://#diff-dc2c87e2a607a99bf28c0758055b605e768efef8f54221ec67b29361540bb2d0R1-R3) [[3]](diffhunk://#diff-4a880866692fbb4e15cc2f0af611ab5502aee10d140728222fb76dc964a57dc4R1-R14) [[4]](diffhunk://#diff-976e53b7515b11188b860c464e190e894f1508a2eced48a22d9ceeeb2ddf3963R1-R24) [[5]](diffhunk://#diff-ac05d954e873baae40c3fb635709eb4779893a852861bedb0c3a40f57df52e97R1-R8) [[6]](diffhunk://#diff-1a3c60779f02cf90e0da3acf8e9b6b3806b85eebf6d31d624cd891b798c31060R1-R3)

### Authentication Updates
* Updated Azure AD configuration across multiple files (`appsettings.json`, `appsettings.Development.json`, `appsettings.Test.json`, and `docker-compose.yaml`) to use new `Domain` and `ClientId` values. [[1]](diffhunk://#diff-d49b123a50e0322cbb86639aaeca1bd7c99c64badae675cf9810bf85917b4632L4-R12) [[2]](diffhunk://#diff-b3c40ae165141f9551789c504c56e827aecaed238db7311715a316b19f62dbffL4-R7) [[3]](diffhunk://#diff-190d0338026aabf08acb92503032bd7dc1a50b700656eb59a1da0afd25ed4e32L4-R12) [[4]](diffhunk://#diff-3fde9d1a396e140fefc7676e1bd237d67b6864552b6f45af1ebcc27bcd0bb6e9L35-R50)
* Changed the claim type used for user identification from `"preferred_username"` to `"oid"` in multiple files, including `AuthenticationExtensions.cs`, `PersonalAccessTokenHandler.cs`, `EmployeeIsActiveHandler.cs`, `RetrieveUsers.cs`, `UserContext.cs`, and associated unit tests. [[1]](diffhunk://#diff-e83cc18008e727fb493576c40b935fc751f298547c7e0e51f5736e6061d1884dL60-R60) [[2]](diffhunk://#diff-cf343461fadaec9c19a5f6db9a11fb90764d4ffc80c99b98786b0149f02fff9fL21-R21) [[3]](diffhunk://#diff-cf343461fadaec9c19a5f6db9a11fb90764d4ffc80c99b98786b0149f02fff9fR30-R34) [[4]](diffhunk://#diff-7715c1c03aef2d8569f9ab1efdc226976d52e74bc64fc36e2591baf2a256c029L23-R23) [[5]](diffhunk://#diff-6aac4e432a28af36082b61c21da2bcd598bcc3990afb20f898e6381500e5aa0dL23-R23) [[6]](diffhunk://#diff-81eb308801cd871105d3cddb32491d3148acfc3386f4fdda9ea54025342cd30eL44-R44) [[7]](diffhunk://#diff-81eb308801cd871105d3cddb32491d3148acfc3386f4fdda9ea54025342cd30eL84-R84)

### Database Seeding Changes
* Updated the email field for a significant user and test user in `MigrationClient.cs` to use a GUID instead of an email address, aligning with the new authentication model. [[1]](diffhunk://#diff-ccb9dbb744e39201492ad00b629fb3fa8b20c1770005e78fce310df8b42303a0L86-R86) [[2]](diffhunk://#diff-ccb9dbb744e39201492ad00b629fb3fa8b20c1770005e78fce310df8b42303a0L252-R253)

### Logging and Debugging
* Changed the default logging level from "Warning" to "Information" in `appsettings.json` and `appsettings.Development.json` to provide more detailed logs during development. [[1]](diffhunk://#diff-d49b123a50e0322cbb86639aaeca1bd7c99c64badae675cf9810bf85917b4632L4-R12) [[2]](diffhunk://#diff-190d0338026aabf08acb92503032bd7dc1a50b700656eb59a1da0afd25ed4e32L4-R12)
* Added additional debugging information to the `EmployeeIsActiveHandler` error message to include user details.

### Minor Fixes and Updates
* Updated the `Dockerfile` for the API to trust HTTPS certificates and removed commented-out lines.
* Added `axios` as a dependency in `package.json` for the new frontend project.

boye-s committed:
Set up basic Vue 3 project. Alvtime 3.0 lives!

boye-s committed:
Try to fix some pathing and linting stuff

boye-s committed:
Workspaces are stupid and useless

boye-s committed:
Implement MSAL Authentication

boye-s committed:
Remove unused initialize method

boye-s committed:
Migrates frontend to v3 with Docker setup

Migrates the frontend application to a new directory structure (v3).

Adds Dockerfile, .dockerignore, and .devcontainer.json for improved
containerization and development environment setup.

Updates the docker-compose file to reflect the new frontend location
and configures port mapping.

boye-s committed:
Updates application styling and fonts

Improves the application's visual appearance by switching to SCSS for styling and incorporating custom fonts from Google Fonts.

This change introduces a more modern and visually appealing design, enhancing the user experience.

boye-s committed:
Configures authentication routes

Updates route configuration to utilize relative imports and add authentication middleware.
Specifically, it configures authentication requirements for the home and secret routes.

boye-s committed:
Sets up basic application layout

Creates a basic application layout with a navigation bar and a router view.

The navigation bar component provides a basic structure for site navigation.
Removes the previous authentication-related code from the App.vue file.

boye-s committed:
Enforces consistent indentation using tabs

Configures ESLint to enforce tab-based indentation for both JavaScript and Vue templates, improving code readability and consistency across the project.

boye-s committed:
Updates path alias for better module resolution

Changes the path alias to allow for more flexible module resolution
within the project.

This simplifies imports and improves code maintainability by
allowing direct references to modules within the `src` directory
without needing the `src` prefix.

boye-s committed:
Loads configuration from JSON file

Loads application configuration from a `config.json` file
instead of hardcoding it. This allows easier configuration
changes without requiring code recompilation.

Updates the authentication configuration to use the loaded
configuration values.

boye-s committed:
Adds absence service

Initializes the absence service to handle API requests for user absence and vacation overviews.

This includes setting up Axios with a base URL and content type, as well as defining functions to retrieve absence and vacation data from the API endpoints.

boye-s committed:
Updates Node.js version in Dockerfile

Updates the base Node.js image version in the Dockerfile
from 20.18 to 22.15. This ensures the application benefits
from the latest features, performance improvements, and
security patches available in the newer Node.js runtime.

boye-s committed:
Sets up global SCSS variables and imports

Establishes a foundation for consistent styling by defining global SCSS variables for colors and fonts.

Configures Vite to automatically import these variables into every component, ensuring a unified look and feel across the application.

boye-s committed:
Adds logout functionality

Implements a logout button and integrates it into the navigation bar.

This allows users to securely sign out of the application, improving
user experience and security.

boye-s committed:
Configures ESLint for the frontend

Sets up ESLint with TypeScript and Vue 3 support, including recommended rules, parser configurations, and ignore patterns.
This ensures code quality and consistency across the frontend codebase.

boye-s committed:
Improves frontend application stability

Addresses several issues to enhance the frontend application's robustness:

- Removes `no-console` rule in production and sets `no-debugger` to warn to prevent issues in production environments.
- Enables the setting of environment variable in public config file.
- Adds error handling within authentication composable to address potential silent request errors.
- Streamlines router guard logic by removing unused parameter.

boye-s committed:
Fixes: Imports missing type specifiers

The original import statement did not explicitly specify the imported types. This change ensures that all imported symbols are correctly recognized as types by adding the `type` keyword to the import statement. This removes potential ambiguity and ensures proper type checking throughout the module.

boye-s committed:
Updates favicon to use ALV logo

Replaces the default Vite favicon with the ALV-specific logo.
This enhances brand recognition and provides a more consistent user experience.

boye-s committed:
Adds application logo

Replaces the text-based logo with an SVG image.

This provides a more visually appealing and brand-consistent experience.
The logo is now displayed in the navigation bar.

boye-s committed:
Removes Hello World component

Removes the default Hello World component, as it's no longer needed.

boye-s committed:
Initializes user store with Pinia

Sets up a Pinia store to manage user data. This includes a reactive user object and a setter method to update the user's information.

boye-s committed:
Ensures user data is set on authentication

This change ensures that the user data is stored in the user store upon successful authentication via redirect. This prevents issues where a user is authenticated but the application does not recognize them, leading to inconsistent behavior.

The user data is retrieved from the MSAL accounts and stored in the user store if it's not already present.

boye-s committed:
Displays the logged-in user's name.

Displays the logged-in user's name in the navigation bar.

This provides visual confirmation of the user's logged-in status.

boye-s committed:
Creates API client with authentication

Initializes an Axios-based API client configured with a base URL and default authorization header.

The authorization token is acquired using MSAL to ensure requests are authenticated with a valid user access token.

boye-s committed:
Imports api client for absence service

The previous implementation defined its own axios instance.
This change imports and utilizes the shared api client,
ensuring consistency and avoiding duplication.

boye-s committed:
Formats logo image tag

Formats the logo image tag in the navigation bar component for improved readability.

boye-s committed:
Adds axios dependency

Adds the axios dependency to the project. This will
allow for making HTTP requests from the frontend.

boye-s committed:
Updates API access scope

Updates the API access scope in the configuration to ensure proper authentication and authorization.

The scope is updated to match the correct application registration.

boye-s committed:
Updates logout button text and styling

Updates the logout button text to Norwegian for better localization.

Adjusts the padding and font size of the logout button for improved visual consistency.

boye-s committed:
Adds base layout for the application

Implements the base layout with a blue bar and content container for the application.

It introduces a wrapper with a blue bar for visual appeal and a content container for the main content with a defined maximum width, padding, and background color.

The background color is moved to the html and body to cover the whole screen.

boye-s committed:
Adds Feather Icons component

Introduces a new `FeatherIcon` component for rendering Feather icons in the application.

This commit adds the `feather-icons` and `vue-feather-icons` dependencies, along with the corresponding type definitions.
The `FeatherIcon.vue` component encapsulates the logic for displaying icons based on their name and size, simplifying their usage throughout the application.
The component also includes basic styling to vertically align the icon with the surrounding text.

boye-s committed:
Adds status bar component

Adds a status bar component to display relevant information, such as data and calendar information.

The status bar enhances the user interface by providing quick access to key metrics.

boye-s committed:
Updates styling and layout of the application.

Refactors the application's visual elements.

Removes the blue bar to create a cleaner interface and modifies the background color for better contrast. Adjusts the navigation bar height to improve visual balance.

boye-s committed:
Replaces home view with status bar

Replaces the placeholder home view with the status bar component.

This provides the user with immediate feedback on the application's status upon opening the application.

boye-s committed:
Re-implements FeatherIcon component

Replaces the FeatherIcon component implementation to utilize Vue's `h` function for rendering. This allows for a more efficient and direct manipulation of the SVG elements, improving performance and control over the icon's attributes.

The previous implementation used `v-html`, which can be less performant and harder to manage. This change improves the component's flexibility and maintainability.

boye-s committed:
Fixes authentication and authorization issues

Updates the authentication scheme to use the 'oid' claim for user identification.

This change addresses issues where user authentication and authorization were failing due to incorrect claim mapping. The 'preferred_username' claim was replaced with 'oid' to align with the actual claim used for user identification. This ensures that user emails are correctly retrieved and used for authorization checks.

Also updates Azure AD ClientId and Domain in docker compose.

boye-s committed:
Updates Azure AD application credentials

Updates the Azure AD application credentials, including the ClientId and Domain, across different environment configurations.

This change ensures the application uses the correct authentication settings.

boye-s committed:
Simplifies Dockerfile for faster builds.

Streamlines the Dockerfile to optimize the build process.

It switches to prebuilt images for faster iteration, removes
unnecessary copying steps, and streamlines the execution
process. It also trusts the dev certificates during build to
allow for HTTPS communication during development.

boye-s committed:
Adds ESLint and Prettier configurations

Sets up ESLint with Vue 3 recommendations and configures Prettier
to enforce consistent code style across the project.

This ensures code quality and readability by automating code formatting and identifying potential issues.

boye-s committed:
Adds secondary and accent colors

Introduces secondary and accent colors to the theme,
providing more flexibility in styling and visual appeal.

boye-s committed:
Updates access scope for user authentication.

Updates the access scope in the configuration file to include profile and openid scopes, enabling enhanced user authentication capabilities. This change is necessary to retrieve user profile information and support OpenID Connect flows.

boye-s committed:
Updates logout button style and adds icon

Enhances the logout button with updated styling and includes a logout icon for improved user experience.
The logout button now has a distinct appearance with a new background color, text color, border radius and padding.

boye-s committed:
Enhances status bar with sun icon and styling

Adds a sun icon to the status bar to display a new value.

Improves the layout and styling of the status bar content boxes
to use flexbox for better spacing and alignment. Adds a minimum
width to prevent the content boxes from collapsing.

boye-s committed:
Fixes typo in absence overview endpoint

Corrects a typo in the `getAbsenceOverview` endpoint URL,
changing `AbsenceOverview` to `AbsenseOverview`.

boye-s committed:
Adds CustomerExpandable component

Creates a reusable component to display customer information,
including name, project, and usage statistics. It also includes
a chevron icon for expand/collapse functionality.

boye-s committed:
Adds TimeTracker component

Adds the TimeTracker component to display the week number and includes a CustomerExpandable component.

This sets up the basic structure for time tracking functionality.

boye-s committed:
Adds TimeTracker component to HomeView

Implements the TimeTracker component in the HomeView
to provide users with a time tracking interface.

Adds basic styling for spacing.

boye-s committed:
Removes unused logo assets

Removes the unused Vite and Vue logo assets from the project.
These files are no longer referenced and contribute to unnecessary
size.

boye-s committed:
Applies code formatting to config.json

Applies code formatting to the config.json file for improved readability and consistency.

boye-s committed:
Renames useUserStore to userStore

Renames the user store file and updates all imports to reflect the change.

This change improves consistency and readability across the codebase by adhering to a more concise naming convention for stores.

boye-s committed:
Includes declaration files in compilation

The change ensures that declaration files are included during the TypeScript compilation process.

This allows the compiler to properly recognize and utilize type declarations from external libraries or modules, preventing potential type-related errors and ensuring type safety across the project.

boye-s committed:
Enforces semicolon usage in JavaScript files.

Adds a rule to the ESLint configuration to enforce the use of semicolons at the end of statements. This change enhances code consistency and readability across the codebase.

boye-s committed:
Adds date and week utility functions

Introduces helper functions for date comparison and manipulation,
including calculating the start of the week and creating a week array
from a given date. This enhances date handling capabilities within
the application.

boye-s committed:
Adds holiday helper

Adds a utility to determine if a given date is a Norwegian holiday.

This includes calculating Easter dates using the `date-easter` package
and providing functions to check for holidays and retrieve holiday
information.

Removes `vue-feather-icons` as it's not longer used.

boye-s committed:
Simplifies todayIsBetween function

Refactors the `todayIsBetween` function to directly use the `new Date()` constructor within the `isBetween` function call, removing the need for an intermediate `today` variable. This improves code readability and conciseness.

boye-s committed:
Creates task service

Creates a service to fetch user tasks from the API.
This provides a dedicated module for handling task-related API interactions.

boye-s committed:
Creates time entries service

Adds a service to fetch time entries for a user within a specified date range.

This service consumes an API endpoint to retrieve time entry data, formatting the dates for the request.

boye-s committed:
Creates time entries store

Initializes a Pinia store to manage time entries.

This store provides functionality to fetch time entries from the backend
using the `timeService` and stores them in a reactive `timeEntries` ref.

It also includes basic error handling for failed API requests.

boye-s committed:
Adds task store with data fetching logic

Implements a Pinia store for managing tasks.

The store fetches tasks from the task service, handles potential errors,
and transforms the data to group tasks by project.

boye-s committed:
Refactors CustomerExpandable to ProjectExpandable

Renames CustomerExpandable component to ProjectExpandable.

Updates the component to display project-specific information,
reflecting the change in naming and functionality.

boye-s committed:
Adds spacing to time tracker projects

Improves the layout of the Time Tracker component by introducing spacing between the project items.

This enhancement makes the project list more readable and visually appealing, creating a better user experience.

boye-s committed:
Enables expandable project view

Implements an expandable section for projects, allowing users to view associated tasks within each project.

Clicking a project expands or collapses the task list.

boye-s committed:
Renders tasks as individual strips

Replaces the basic task display in the project expandable with a new TaskStrip component.

This change provides a more structured and presentable view of the tasks associated with a project, and displays task name and its compensation rate.

boye-s committed:
Removes unnecessary file type from tsconfig

Removes the `tsx` file type from the `include` array in the `tsconfig.app.json` file.

This change improves build performance by preventing the compiler from processing files that are not needed for the application build, as the project primarily uses `vue` files for components.

boye-s committed:
Refactors: Moves project type definitions

Centralizes project, customer, and task type definitions in a dedicated file.

This change promotes code reusability and maintainability by defining these shared types in a single location.
It removes duplicate interface definitions from individual components, ensuring consistency across the application.

boye-s committed:
Replaces Feather icons with Hugeicons in status bar

Migrates from Feather icons to Hugeicons in the status bar component for improved icon library and visual consistency.

Updates the status bar component to use Hugeicons and removes the dependency on Feather icons.

boye-s committed:
Adds a debounce function

Implements a debounce function to limit the rate at which a function is executed.

This prevents excessive calls to the function, improving performance.

boye-s committed:
Adds time entry update functionality

Introduces the ability to update time entries via a new API endpoint.

This allows users to modify existing time entries,
enabling corrections and adjustments to their recorded work hours.

A new type definition for TimeEntry is added to ensure
type safety when dealing with time entry data.

boye-s committed:
Adds swiper dependency

Adds the swiper library to the project.

This enables the implementation of interactive
and responsive image carousels and sliders.

boye-s committed:
Improves debounce function with context binding

Enhances the debounce function to properly maintain the context (`this`)
when the debounced function is executed. This ensures that the function
behaves as expected, especially within class methods or event handlers.
Also makes the delay a default parameter.

boye-s committed:
Adds hour input component

Creates a reusable hour input component for time tracking.

This component allows users to input time values, handles locale-specific formatting (Norwegian), and updates the time entry in the store.

The component includes basic styling and utilizes Vue's computed properties for two-way data binding.

boye-s committed:
Creates TaskWeek component

Adds TaskWeek component to display hour inputs for each day of the week for a given task.

Ensures a time entry exists for each day in the week, creating a new entry with default values if one doesn't already exist.

Sorts time entries by date.

boye-s committed:
Adds TaskStrip component

Creates a component to display task information and its associated week.

This component includes the task name, compensation rate, and a child component to manage task details for a specific week.

Styles the component to improve layout and visual appeal.

boye-s committed:
Passes week prop to TaskStrip component

The week prop is now passed to the TaskStrip component. This allows the TaskStrip component to access and utilize the week information, which is presumably needed for task-related logic or display within that component.

boye-s committed:
Introduces date management with Pinia stores

Adds Pinia stores for managing the active date, weeks, and time entries.
This includes creating stores for `dateStore` and `timeEntriesStore` to manage date state and time entry data, respectively.

Refactors `timeService` to accept parameters as an object.

Improves time entry handling with queueing and debouncing for efficient updates.

boye-s committed:
Configures Vue to enforce attribute hyphenation

Updates the Vue ESLint configuration to enforce attribute hyphenation, improving code consistency and readability.

boye-s committed:
Includes `.tsx` files in TypeScript compilation

The TypeScript configuration now includes `.tsx` files
for compilation. This ensures that TypeScript processes
and transpiles these files along with `.ts` files.

boye-s committed:
Improves Time Tracker UI spacing and readability

Increases the gap between elements in the project container for better visual separation.

Enhances the font size of the task strip for improved readability.

boye-s committed:
Adds eslint disable for explicit any

Adds an eslint disable comment to allow the use of 'any'
type in a specific scenario. This is done to address a
temporary situation where a more specific type is not yet
available or practical, while acknowledging the potential
risks and the need for future refinement.

boye-s committed:
Registers Swiper custom elements

Registers Swiper custom elements for use in the application.

This allows the use of Swiper elements directly in the application's templates.

boye-s committed:
Implements week-based time tracking

Introduces a week-based time tracking feature using a swiper component for navigating between weeks.

This allows users to view and manage their tracked time on a weekly basis.
It also integrates with the date store to retrieve week data and utilizes a utility function to display week numbers.

boye-s committed:
Adds ID to hour input

Adds an ID attribute to the hour input field in the Time Tracker component.

This ID is constructed using the task ID and date, which allows for easier identification and manipulation of the input field in automated tests and other contexts.

boye-s committed:
Do not try to render web components as Vue components

boye-s committed:
Awaits time entries retrieval

Ensures that time entries are fetched before proceeding,
preventing potential race conditions.

boye-s committed:
Improves TimeTracker component navigation

Adds navigation buttons ("Tilbake", "I dag", "Fremover") to the
TimeTracker component, allowing users to move between weeks.
Also formats the week number display to include the year when
displaying weeks from different years, and initializes the Swiper
component to start on the current week.

boye-s committed:
Delays TimeTracker load until data is fetched

Ensures that the TimeTracker component only renders after the necessary data (active date and tasks) has been fetched.

This prevents potential issues or errors that might arise from the component attempting to access data before it's available.

boye-s committed:
Improves time tracker navigation and UI

Enhances the time tracker component by adding navigation buttons with icons and a "Today" button to quickly jump to the current week.

It also adjusts the hover color of the logout button for better visual consistency and calculates the initial slide index for the week swiper.

Updates styling for buttons in TimeTracker to have rounded corners and consistent padding.

boye-s committed:
Adds day pill component for time tracking

Adds a DayPill component to display a date in a pill format.
Also introduces a DayPillStrip component to display a row of DayPills for the week.
The dayOfWeek utility function is added to format the day name.

boye-s committed:
Formats time value and adjusts input width

Formats the time value using the Norwegian locale ("nb-NO")
to ensure consistent decimal formatting.

Increases the width of the input field to accommodate the
formatted time value, improving usability.

boye-s committed:
Adds day pill strip to project expandable

Adds a day pill strip to the project expandable component.

This allows users to quickly see which days have time entries for a specific project.

boye-s committed:
Adds padding to page content

Improves the layout by adding padding to the bottom of the wrapper and adding margin to the content container.

This change ensures that the content is spaced appropriately from the bottom of the screen.

boye-s committed:
Improves day pill and project expandable UI

Enhances the DayPill component to display holiday descriptions on hover and adds a visual distinction for holidays.

Updates ProjectExpandable to use task store for controlling project expansion state.
This resolves an issue where the expandable state was not being properly maintained across components.

boye-s committed:
Highlights time entry input when filled

Improves user experience by visually highlighting the
time entry input field when a value greater than 0 is entered.

This provides immediate feedback to the user, indicating
that the field has been successfully populated with a value.

boye-s committed:
Adds project expandable toggle

Adds a function to toggle the expandable state of a project within the task list.

This allows users to collapse or expand project details for better task management.
The `open` property is initialized to `false` when a new project is created.

boye-s committed:
Corrects holiday descriptions

Updates the spelling and formatting of Norwegian holiday descriptions
for improved readability.

boye-s committed:
Adds project open state

Adds an `open` field to the `Project` type definition.

This field indicates whether a project is currently active or closed,
allowing for better project management and filtering.

boye-s committed:
Populates time entries on task week mount

Ensures all days of the week have a time entry,
either existing or new, when the TaskWeek component mounts.
This allows users to easily input time for each day
without manually creating entries.

boye-s committed:
Adds "Track rest of day" button to time entries

Implements a button that allows users to quickly fill in the remaining time in their workday for a specific task.
This improves the user experience by providing a convenient way to track time.
The button dynamically calculates the remaining time based on existing time entries.

boye-s committed:
Formats TimeTracker component

Addresses minor formatting inconsistencies within the TimeTracker component in the HomeView.

boye-s committed:
Fixes time entry update and date handling

Updates the time entry update logic to correctly handle localized number formats and prevent data loss.

Ensures correct date representation across timezones when creating new task entries for a week. This prevents issues where entries might be associated with the wrong date.

Refactors the time value handling in the `HourInput` component for improved accuracy and user experience.

boye-s committed:
Refactors ProjectExpandable component to use slots

Moves the project stats and task list into slots within the
ProjectExpandable component.

This change improves flexibility and allows for easier customization
of the component's content. Removes the week property, DayPillStrip
and TaskStrip from the component and passes them through the slot in
the TimeTracker component.

Also uses a chevron-up icon when the project is open and chevron-down
when it's closed.

boye-s committed:
Adds task activity management

Introduces task activity management feature, enabling users to view and interact with project tasks.

This includes:
- Displaying project tasks with name and completion rate.
- Allowing users to toggle comments and favorite status on tasks.
- Persisting changes to local storage.

Removes the secret view and adds the task view in the router.

boye-s committed:
Updates navigation bar with links

Improves the navigation bar by adding links to different sections of the application.
The navigation bar now includes links for Timeføring (Time Tracking) and Aktiviteter (Activities).
It also introduces a user container to display the user's name and the logout button.
Styling updates enhance the overall look and feel.

boye-s committed:
Removes console log statement

Removes a console log statement that was likely used for debugging purposes.

boye-s committed:
Adds comment functionality to time entries

Adds components for displaying and editing comments associated with time entries.

The `CommentPill` component provides a visual indicator for entries with comments.

The `TimeEntryComment` component enables users to add and update comments with a debounced update mechanism. Also shows a timestamp of the latest update.

boye-s committed:
Adds time entry comment functionality

Implements the ability to add and update comments on time entries.

This enhances the time tracking feature by allowing users to provide
additional context and details about their work hours.
It introduces a comment pill and a comment input component that
are displayed conditionally based on user interaction and settings.

boye-s committed:
Adds comment enabling to HourInput

Refactors TaskWeek to directly use filtered time entries from the store.

This change removes the local time entries model and computes
`filteredTimeEntriesForTask` directly, improving reactivity.

It also passes the task's `enableComments` property to the
HourInput component, enabling or disabling commenting functionality
based on the task settings.

boye-s committed:
Adds frontend-v3 service to docker-compose

Adds a new service definition for the frontend-v3
application within the docker-compose configuration.
This facilitates local development and testing.

palbock commented on packages/api/AlvTime.Persistence/MigrationClient.cs:
```diff
@@ -83,7 +83,7 @@ private static async Task SeedAccessTokens(AlvTime_dbContext context)
         if (!accessTokens.Any())
         {
-            var significantUser = context.User.First(u => u.Email == "ahre-ketil.lillehagen@alvno.onmicrosoft.com");
+            var significantUser = context.User.First(u => u.Email == "e6dd42da-3cfc-4c39-aee5-6868aad184fe");
```
Tenker vi heller burde bruke u.Oid istedet for Email

palbock commented on packages/api/AlvTime.Persistence/MigrationClient.cs:
```diff
@@ -249,8 +249,8 @@ private static async Task SeedUsers(AlvTime_dbContext context)
-                    Name = "Ahre Ketil Lillehagen",
-                    Email = "ahre-ketil.lillehagen@alvno.onmicrosoft.com",
+                    Name = "Boye Valentin Nerli Svensson",
+                    Email = "e6dd42da-3cfc-4c39-aee5-6868aad184fe",
```
Oid

palbock commented on packages/api/AlvTimeWebApi/Authentication/PersonalAccessToken/PersonalAccessTokenHandler.cs:
```diff
@@ -57,7 +57,7 @@ private static AuthenticationTicket CreateTicket(IEnumerable<Claim> claims)
             => new[]
             {
-                new Claim("preferred_username", user.Email),
+                new Claim("oid", user.Email),
```
user.Oid

palbock commented on packages/api/AlvTimeWebApi/Authorization/Handlers/EmployeeIsActiveHandler.cs:
```diff
@@ -18,7 +18,7 @@ public EmployeeIsActiveHandler(AlvTime_dbContext alvtimeDbContext)
     protected override Task HandleRequirementAsync(AuthorizationHandlerContext authContext, EmployeeStillActiveRequirement requirement)
     {
-        var userEmail = authContext.User.Claims.FirstOrDefault(c => c.Type == "preferred_username")?.Value;
+        var userEmail = authContext.User.Claims.FirstOrDefault(c => c.Type == "oid")?.Value;
```
Jeg tror at å slå opp på "oid" kun vil fungere for PAT siden du eksplisitt kaller den for "oid" i handleren der, men i tokenet som kommer fra Azure heter claimet http://schemas.microsoft.com/identity/claims/objectidentifier, så vi burde nok bruke det overalt.

palbock commented on packages/api/AlvTimeWebApi/Controllers/Utils/RetrieveUsers.cs:
```diff
@@ -20,7 +20,7 @@ public RetrieveUsers(AlvTime_dbContext database, IHttpContextAccessor httpContex
         public User RetrieveUser()
         {
-            var email = _httpContextAccessor.HttpContext?.User.FindFirstValue("preferred_username");
+            var email = _httpContextAccessor.HttpContext?.User.FindFirstValue("oid");
```
oid -> http://schemas.microsoft.com/identity/claims/objectidentifier

palbock commented on packages/api/AlvTimeWebApi/Controllers/Utils/UserContext.cs:
```diff
@@ -20,7 +20,7 @@ public UserContext(IHttpContextAccessor httpContextAccessor, IUserRepository use
         private string Name => _httpContextAccessor.HttpContext.User.FindFirstValue(ClaimTypes.Name);
 
-        private string Email => _httpContextAccessor.HttpContext.User.FindFirstValue("preferred_username");
+        private string Email => _httpContextAccessor.HttpContext.User.FindFirstValue("oid");
```
oid -> http://schemas.microsoft.com/identity/claims/objectidentifier

Commit:
La til .env i gitignore

Commit:
Endret småting fra rebase

Commit:
Forenklet Dockerfile og la tilbake Ahre-Ketil i seeding

boye-s committed:
Updates application client ID

Updates the application's client ID and access scope
in the configuration file.

This change ensures the application uses the correct
credentials for authentication and authorization.

boye-s committed:
Updates significant user identifier

Updates the hardcoded user identifier used for initial access token generation and employee ID assignment.

This change ensures that the application uses the correct user identifier, resolving potential authentication and data association issues related to the original hardcoded value.

boye-s committed:
Enables top-level await in browser builds.

Configures the build process to support top-level await features,
ensuring compatibility with modern browsers that support this
functionality.
